tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects_test.ts

```diff
@@ -205,283 +230,433 @@ describe('Debugger effects', () => {
+            stack_frames: [stackFrame0, stackFrame1],
+          });
+
+          action.next(
+            executionDigestFocused({displayIndex: displayIndexOfFocus})
+          );
+
+          if (dataAlreadyExists) {
+            // Data already exists.
+            expect(fetchExecutionData).not.toHaveBeenCalled();
+            expect(fetchStackFrames).not.toHaveBeenCalled();
+          } else {
+            expect(fetchExecutionData).toHaveBeenCalledTimes(1);
+            expect(fetchStackFrames).toHaveBeenCalledTimes(1);
+            expect(dispatchSpy).toHaveBeenCalledWith(
+              executionDataLoaded(executionDataResponse)
+            );
+            expect(dispatchSpy).toHaveBeenCalledWith(
+              stackFramesLoaded({
+                stackFrames: {aa: stackFrame0, bb: stackFrame1},
+              })
+            );
+          }
+        }
+      );
+    }
 
-    for (const triggeringAction of [
-      executionScrollLeft(),
-      executionScrollRight(),
-    ]) {
+    for (const dataAlreadyExists of [false, true]) {
       it(
-        `Scrolling ${triggeringAction.type} ` +
-          ` leads to execution-digest request`,
+        `scrolling right triggers execution digest loading: ` +
+          `dataAlreadyExists=${dataAlreadyExists}`,
         () => {
-          store.setState(
-            createState(
-              createDebuggerState({
-                activeRunId: '__default_debugger_run__',
-                executions: {
-                  numExecutionsLoaded: {
-                    state: DataLoadState.LOADED,
-                    lastLoadedTimeInMs: 1234,
-                  },
-                  executionDigestsLoaded: {
-                    state: DataLoadState.LOADED,
-                    lastLoadedTimeInMs: 5678,
-                    numExecutions: 6,
-                    pageLoadedSizes: {0: 2},
-                  },
-                  pageSize: 2,
-                  displayCount: 2,
-                  scrollBeginIndex: 2,
-                  executionDigests: {},
-                },
-              })
-            )
+          const dataAlreadyExists = false;
+          const originalScrollBeginIndex = 50;
+          const scrollBeginIndex = originalScrollBeginIndex + 1;
+          const numExecutions = 100;
+          const displayCount = 10;
+          const pageSize = 20;
+          store.overrideSelector(getActiveRunId, runId);
+          store.overrideSelector(
+            getExecutionScrollBeginIndex,
+            scrollBeginIndex
+          );
+          store.overrideSelector(getNumExecutions, numExecutions);
+          store.overrideSelector(getDisplayCount, displayCount);
+          store.overrideSelector(getExecutionPageSize, pageSize);
+          let pageLoadedSizes: {[pageIndex: number]: number} = {};
+          pageLoadedSizes = {
+            0: 20,
+            1: 20,
+            2: 20,
+          };
+          if (dataAlreadyExists) {
+            pageLoadedSizes[3] = 5;
+          }
```

Comment:
1. Done.
2. I prefer to keep the current style. The `for-of` part currently states clearly what condition is being tested (e.g., whether data already exists or not). If the `for-of` iterates over numbers instead of booleans, it'll be harder to follow what the tests are about. I agree the logic inside test code shouldn't be overly complex. But I don't think I've cross the line here yet.